resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala

```diff
@@ -143,6 +144,9 @@ private[spark] class Client(
         logError("Please check \"kubectl auth can-i create [resource]\" first." +
           " It should be yes. And please also check your feature step implementation.")
         kubernetesClient.resourceList(preKubernetesResources: _*).delete()
+        // register shutdownhook for cleaning up the upload dir only
```

Comment:
This is a weird place to register a shutdown hook, why not just register earlier instead of inside all of these exception cases?

Reply:
The reason we need to register only in exception cases is that the cleanup of files and directory under file upload path should happen on client side only if job submission fails. If we register early, and job is submitted with 
spark.kubernetes.submission.waitAppCompletion conf value as false, shutdown hook will be called as soon as app submission completes and the files will be deleted. This will case the job to fail as driver will not be able to find these files.

Reply:
@holdenk do you have any suggestion to handle this in a better way?

Reply:
ah interesting. What about if we only register in client mode?

Reply:
Client mode will not help as in that case cleanup will be missed if job submission fails for cluster deploy mode.
We are doing some other resource cleanup activities in the catch blocks of this class. So, registering the upload path cleanup should also be ok.